### ISSUE TITLE
parallel wr.s3.to_parquet() writes and glue catalog table updates

### ISSUE THREAD

**Issue by ghukill** [awslabs/aws-data-wrangler#526]
**UPDATE:** fwiw, I wholly understand if this question is outside the scope of this forum, as it's more of a situational context with awswrangler functionality that is operating as expected.  But figured it couldn't hurt to raise.

----------------------

I am encountering a tricky situation when attempting to run `wr.s3.to_parquet()` in parallel - for different dataframes -- that are writing to the same parquet dataset (different partitions), but all updating the same glue catalog table.

The specific problem I'm facing: not all columns from written partitions are present in glue catalog table.

Context: running an AWS StepFunction with lambdas executing in parellel to write dataframes to parquet dataset in S3.  

Becuase the dataframes are always unique partitions, that works splendidly for the actual parquet file writes to S3.  The problem I'm encountering is when parallel processes attempt to update the **glue catalog table** -- call it `foo` -- at nearly the same time, via the `wr.s3.to_parquet()` function.

Digging into that function, I can see where it a) retrieves the current glue catalog input for the table and b) when it attempts to update/create that table:

```python
# subset from wr.s3.to_parquet()
...
# Evaluating dtype
    catalog_table_input: Optional[Dict[str, Any]] = None
    if database is not None and table is not None:
        catalog_table_input = catalog._get_table_input(  # pylint: disable=protected-access    <------ a) get catalog table
            database=database, table=table, boto3_session=session, catalog_id=catalog_id
        )
...
...
...
        if (database is not None) and (table is not None):
            try:
                catalog._create_parquet_table(  # pylint: disable=protected-access   <------- b) create/update catalog table
                    database=database,
                    table=table,
                    path=path,
                    columns_types=columns_types,
                    partitions_types=partitions_types,
                    compression=compression,
                    description=description,
                    parameters=parameters,
                    columns_comments=columns_comments,
                    boto3_session=session,
                    mode=mode,
                    catalog_versioning=catalog_versioning,
                    projection_enabled=projection_enabled,
                    projection_types=projection_types,
                    projection_ranges=projection_ranges,
                    projection_values=projection_values,
                    projection_intervals=projection_intervals,
                    projection_digits=projection_digits,
                    catalog_id=catalog_id,
                    catalog_table_input=catalog_table_input,   <---------------- passed catalog table input
                )              
```

I'm still confirming this is precisely the issue, but it makes sense that it could be.  I wondered if there were any recommendations for how to think about approaching this?  

**Possible approach:** before *any* of the parallel writes, aggregate all columns from all dataframes (which I can do in this situation) and create the glue catalog table in advance.  Believe I could reuse some awswrangler functions and methods to get those handy `dtype` pandas --> pyarrow --> athena conversions.  

Is that the best bet here?  Or had wondered if there could be some versioning catalog tables and merging of versions after all the writes?

Thanks for a really wonderful library; despite unique situation/context, it's been a joy to work with and really streamlining our work.

**Comment by igorborgest** [comment]
Hi @ghukill, thanks for reaching out!

Actually this is a pretty interesting use case, let me explain my point of view.

If the catalog would have the concept of transaction/lock, we would be able to overcome this situation naturally.  Unfortunately it hasn't, so we would not be able to ensure consistency during multiple parallel "updates" w/o a more advanced/complex strategy.

## Scenario 1 - You know the final schema

This is the simplest case where I would resolve creating the table first and than write passing `schema_evolution=False`.

e.g.
```py3
import awswrangler as wr
import pandas as pd

DATABASE = "default"
TABLE = "test"
PATH = f"s3://BUCKET/{TABLE}/"

wr.s3.delete_objects(PATH)
wr.catalog.create_parquet_table(
    database=DATABASE,
    table=TABLE,
    path=PATH,
    columns_types={"c0": "string", "c1": "double", "c2": "bigint"},
    partitions_types={"p0": "int"},
)

for df in (  # Parallel simulation
    pd.DataFrame({"c0": ["foo"], "p0": [1]}),
    pd.DataFrame({"c1": [0.0], "p0": [2]}),
    pd.DataFrame({"c2": [0], "p0": [3]}),
):
    wr.s3.to_parquet(
        df,
        PATH,
        dataset=True,
        mode="append",
        table=TABLE,
        database=DATABASE,
        partition_cols=["p0"],
        schema_evolution=False,
    )
```

## Scenario 2 - You don't know the final schema but the catalog can be updated afterwards

You can ignore any catalog change during the write and call `wr.s3.store_parquet_metadata()` in the end.

e.g.
```py3
import awswrangler as wr
import pandas as pd

DATABASE = "default"
TABLE = "test"
PATH = f"s3://BUCKET/{TABLE}/"

wr.s3.delete_objects(PATH)

for df in (  # Parallel simulation
    pd.DataFrame({"c0": ["foo"], "p0": [1]}),
    pd.DataFrame({"c1": [0.0], "p0": [2]}),
    pd.DataFrame({"c2": [0], "p0": [3]}),
):
    wr.s3.to_parquet(
        df,
        PATH,
        dataset=True,
        mode="append",
        partition_cols=["p0"]
    )

wr.s3.store_parquet_metadata(
        PATH,
        dataset=True,
        database=DATABASE,
        table=TABLE,
)
```

## Scenario 3 - You don't know the final schema and the catalog can't be updated afterwards

This is the hardest case, and I think the only solution would be keep/manage your own lock using something like S3, DynamoDB, Redis (Thinking in a distributed situation where to keep it in memory is not an option). With this lock mechanism you would be able to run your writes "sequentially" keeping the rest of the pipeline (reading, processing, etc) parallel.

e.g.
```py3
import awswrangler as wr
import pandas as pd

DATABASE = "default"
TABLE = "test"
PATH = f"s3://BUCKET/{TABLE}/"

wr.s3.delete_objects(PATH)
wr.catalog.create_parquet_table(
    database=DATABASE,
    table=TABLE,
    path=PATH,
    columns_types={"c0": "string", "c1": "double", "c2": "bigint"},
    partitions_types={"p0": "int"},
)

for df in (  # Parallel simulation
    pd.DataFrame({"c0": ["foo"], "p0": [1]}),
    pd.DataFrame({"c1": [0.0], "p0": [2]}),
    pd.DataFrame({"c2": [0], "p0": [3]}),
):

    # READ AND PROCESS ANYTHING YOU WANT HERE IN PARALLEL

    # GET/WAIT THE LOCK
    wr.s3.to_parquet(
        df,
        PATH,
        dataset=True,
        mode="append",
        table=TABLE,
        database=DATABASE,
        partition_cols=["p0"],
        schema_evolution=False,
    )
    # RELEASE THE LOCK

```

---

What do you think? Any other ideas?

**Comment by ghukill** [comment]
Response: https://github.com/awslabs/aws-data-wrangler/issues/526

----------------------------------------------------

Thanks for the detailed response and suggestions!  Much appreciated.

In the interim, went strongly in the direction of **Scenario 1**, knowing the final schema in advance and updating Glue catalog before parallel writes occur.  This required tightening up the data typing for the writes in general, so it was fortuitous.

However, the actual implementation varied a bit.  In my situation, the Glue table may or may not exist, and the particular set of parallel writing "handlers" may vary.

The deviation was to:

  * create a dataframe that includes all columns for all handlers, known in advance
  * insert a single row with NULL values across
  * type each row from the **pandas** type we also -- conveniently -- know for that column
  * use `wr.s3.to_parquet()` to write the dataframe, including `table` and `database` so that Glue table is created/updated
  * immediately delete the temporary write (single partition with single, empty row sans two parition columns)

I'm re-remembering as I type this out that this was primarily to utilize the very useful pandas --> pyarrow --> Athena conversions that I believe `wr.s3.to_parquet()` performs?  By writing a dataframe with the same `dtypes` that we'll write columnar subsets of later, seemed to ensure it would all align nicely.  And it does!  

Another wrinkle: these parallel writes are occurring in waves, where 2nd, 3rd, 4th wave handlers may potentially perform Athena queries against the table in question to read data that upstream handlers wrote.  This cycle of dataframe --> S3 --> Glue --> Athena --> dataframe is precisely why this library has been such a productivity boost.

In that moment between waves, though, we have opportunity to pause and update the Glue table sequentially, making **Scenario 2** also intriguing.  The only hurdle I see there, we'd need to know the **Glue/Athena** data types for the columns it looks like, versus the **pandas** data types we _do_ know.  Which was part of the thinking on relying on the empty dataframe write to leverage those mechanics.  I poked around a bit, and saw some some internal methods, but is there a convenient way to get Glue/Athena data types from a pandas dataframe?

Scenario 2 is also appealing because it would allow _not_ updating the Glue table in situations where the handlers fail.  Scenario 1 does not readily support this.

In summary: not sure!  The project is still in its infancy, so maybe there are requirements I'm not yet seeing, but this has been really helpful.  Feel free to close out this question if you'd like, and many thanks again.

**Comment by igorborgest** [comment]
Definitively you have a cool puzzle to solve. Of course I will not be able to point the best direction for you specific case.

But here are two points that might help:

- We have [wr.catalog.extract_athena_types()](https://aws-data-wrangler.readthedocs.io/en/stable/stubs/awswrangler.catalog.extract_athena_types.html) that can help to infer the equivalent Athena types from a Pandas DataFrame.
- It is possible to use [wr.s3.store_parquet_metadata()](https://aws-data-wrangler.readthedocs.io/en/stable/stubs/awswrangler.s3.store_parquet_metadata.html) with the argument `mode="append"` to update your table evolving your schema instead of overwrite it.

**Comment by ghukill** [comment]
And there it is: `wr.catalog.extract_athena_types()`, I must have sailed right past it.  Thanks for pointing this out!
